safe_relay_service/relay/tasks.py

```diff
@@ -395,3 +381,109 @@ def find_erc_20_721_transfers_task() -> int:
     except LockError:
         pass
     return number_safes
+
+
+@app.shared_task(bind=True, soft_time_limit=LOCK_TIMEOUT, max_retries=3)
+def begin_circles_onboarding_task(self, safe_address: str) -> None:
+    """
+    Starts a multi-step onboarding task for Circles users which 1. funds
+    deploys a Gnosis Safe for them 2. funds the deployment of their Token.
+    :param safe_address: Address of the safe to-be-created
+    """
+
+    assert check_checksum(safe_address)
+
+    redis = RedisRepository().redis
+    lock_name = f'locks:begin_circles_onboarding_task:{safe_address}'
+    try:
+        with redis.lock(lock_name, blocking_timeout=1, timeout=LOCK_TIMEOUT):
+            logger.info('Start onboarding for Circles Safe {}'.format(safe_address))
+            # Deploy Safe when it does not exist yet
+            safe_creation2 = SafeCreation2.objects.get(safe=safe_address)
+            if not safe_creation2.tx_hash:
+                logger.info('Safe does not exist yet, start deploying it {}'.format(safe_address))
+                circles_onboarding_safe_task.delay(safe_address)
+                # Retry later to check for Token funding
+                raise self.retry(countdown=30)
+            else:
+                logger.info('Safe exists, start funding Token {}'.format(safe_address))
+                # Fund Token deployment when it does not exist yet
+                circles_onboarding_token_task.delay(safe_address)
+    except LockError:
+        pass
+
+
+@app.shared_task(soft_time_limit=LOCK_TIMEOUT, max_retries=3)
+def circles_onboarding_safe_task(safe_address: str) -> None:
```

Comment:
if my understanding is right, this task usually has to run twice to actually deploy the safe - once to fund it and then once to deploy. if this has to run twice, I'm concerned the 30 sec delay above may not be enough to guarantee this is finished on a real network with 10 second blocktime ...